File: tests/pnl-logger.test.ts
import path from "path";
import fs from "fs";

// Use a temp DB for tests
process.env["DB_PATH_OVERRIDE"] = path.join(__dirname, "test-trades.db");

// We need to patch the DB path before importing pnl-logger
// Since better-sqlite3 creates the file on open, we just make sure cleanup happens.
const TEST_DB = path.join(__dirname, "test-trades.db");

afterAll(() => {
  if (fs.existsSync(TEST_DB)) {
    fs.unlinkSync(TEST_DB);
  }
});

// Reimport after env setup
import { logTrade, getDailyStats, getRecentTrades, getLifetimePnL, closeDb } from "../src/pnl-logger";
import { ArbitrageOpportunity, TradeResult } from "../src/types";

const mockOpportunity: ArbitrageOpportunity = {
  pair: "SOL/USDC",
  buy_dex: "raydium",
  buy_price: 79.28,
  sell_dex: "orca",
  sell_price: 79.45,
  spread_pct: 0.21,
  estimated_profit_usd: 42.5,
  confidence: 0.85,
};

// IDs are generated fresh each test run to avoid UNIQUE constraint failures
function makeSuccessResult(): TradeResult {
  return {
    trade_id: `test-${Date.now()}-success`,
    status: "success",
    buy_tx: "txhash1",
    sell_tx: "txhash2",
    profit_usd: 38.2,
    fees_usd: 4.3,
    net_pnl_usd: 33.9,
    timestamp: new Date().toISOString(),
  };
}

function makeFailedResult(): TradeResult {
  return {
    trade_id: `test-${Date.now()}-failed`,
    status: "failed",
    buy_tx: null,
    sell_tx: null,
    profit_usd: 0,
    fees_usd: 0,
    net_pnl_usd: 0,
    timestamp: new Date().toISOString(),
    error: "Slippage exceeded",
  };
}

describe("PnL Logger", () => {
  beforeAll(() => {
    // Seed two trades so stats queries have data to work with
    logTrade(mockOpportunity, makeSuccessResult(), 100);
    logTrade(mockOpportunity, makeFailedResult(), 100);
  });

  afterAll(() => {
    closeDb();
  });

  it("logs a successful trade without throwing", () => {
    expect(() => logTrade(mockOpportunity, makeSuccessResult(), 100)).not.toThrow();
  });

  it("logs a failed trade without throwing", () => {
    expect(() => logTrade(mockOpportunity, makeFailedResult(), 100)).not.toThrow();
  });

  it("getDailyStats returns correct counts and PnL", () => {
    const today = new Date().toISOString().split("T")[0];
    const stats = getDailyStats(today);

    expect(stats.trade_count).toBeGreaterThanOrEqual(2);
    expect(stats.win_rate).toBeGreaterThanOrEqual(0);
    expect(stats.win_rate).toBeLessThanOrEqual(100);
    expect(typeof stats.total_pnl).toBe("number");
  });

  it("getRecentTrades returns an array", () => {
    const trades = getRecentTrades(5);
    expect(Array.isArray(trades)).toBe(true);
  });

  it("getLifetimePnL returns a number", () => {
    const pnl = getLifetimePnL();
    expect(typeof pnl).toBe("number");
  });

  it("tx_hashes are parsed back as arrays", () => {
    const trades = getRecentTrades(10);
    for (const trade of trades) {
      expect(Array.isArray(trade.tx_hashes)).toBe(true);
    }
  });
});

File: go/internal/pnllogger/pnl_logger_test.go
package pnllogger_test

import (
	"fmt"
	"os"
	"path/filepath"
	"testing"
	"time"

	"github.com/steuf11/arb-bot/internal/pnllogger"
	"github.com/steuf11/arb-bot/internal/types"
)

func TestMain(m *testing.M) {
	// Use a temp DB so tests don't touch the real trades.db.
	tmp, err := os.MkdirTemp("", "arb-bot-test-*")
	if err != nil {
		fmt.Fprintf(os.Stderr, "failed to create temp dir: %v\n", err)
		os.Exit(1)
	}
	os.Setenv("DB_PATH_OVERRIDE", filepath.Join(tmp, "test-trades.db"))

	code := m.Run()

	pnllogger.CloseDB()
	os.RemoveAll(tmp)
	os.Exit(code)
}

var mockOpportunity = types.ArbitrageOpportunity{
	Pair:               "SOL/USDC",
	BuyDex:             "raydium",
	BuyPrice:           79.28,
	SellDex:            "orca",
	SellPrice:          79.45,
	SpreadPct:          0.21,
	EstimatedProfitUSD: 42.5,
	Confidence:         0.85,
}

func newSuccessResult() types.TradeResult {
	buyTx := "txhash1"
	sellTx := "txhash2"
	return types.TradeResult{
		TradeID:   fmt.Sprintf("test-%d-success", time.Now().UnixNano()),
		Status:    types.TradeSuccess,
		BuyTx:     &buyTx,
		SellTx:    &sellTx,
		ProfitUSD: 38.2,
		FeesUSD:   4.3,
		NetPnLUSD: 33.9,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}
}

func newFailedResult() types.TradeResult {
	errMsg := "Slippage exceeded"
	return types.TradeResult{
		TradeID:   fmt.Sprintf("test-%d-failed", time.Now().UnixNano()),
		Status:    types.TradeFailed,
		ProfitUSD: 0,
		FeesUSD:   0,
		NetPnLUSD: 0,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Error:     &errMsg,
	}
}

func TestLogSuccessfulTrade(t *testing.T) {
	if err := pnllogger.LogTrade(mockOpportunity, newSuccessResult(), 100); err != nil {
		t.Errorf("LogTrade (success) returned error: %v", err)
	}
}

func TestLogFailedTrade(t *testing.T) {
	if err := pnllogger.LogTrade(mockOpportunity, newFailedResult(), 100); err != nil {
		t.Errorf("LogTrade (failed) returned error: %v", err)
	}
}

func TestGetDailyStatsCountsAndPnL(t *testing.T) {
	// Seed at least two trades for today.
	_ = pnllogger.LogTrade(mockOpportunity, newSuccessResult(), 100)
	_ = pnllogger.LogTrade(mockOpportunity, newFailedResult(), 100)

	stats, err := pnllogger.GetDailyStats("")
	if err != nil {
		t.Fatalf("GetDailyStats: %v", err)
	}
	if stats.TradeCount < 2 {
		t.Errorf("expected trade_count >= 2, got %d", stats.TradeCount)
	}
	if stats.WinRate < 0 || stats.WinRate > 100 {
		t.Errorf("win_rate out of range: %.1f", stats.WinRate)
	}
}

func TestGetRecentTradesReturnsSlice(t *testing.T) {
	trades, err := pnllogger.GetRecentTrades(5)
	if err != nil {
		t.Fatalf("GetRecentTrades: %v", err)
	}
	// trades is a slice (possibly empty) — just verify the type is correct.
	_ = trades
}

func TestGetLifetimePnLReturnsNumber(t *testing.T) {
	pnl, err := pnllogger.GetLifetimePnL()
	if err != nil {
		t.Fatalf("GetLifetimePnL: %v", err)
	}
	// Must be a finite float — no NaN/Inf.
	if pnl != pnl {
		t.Error("GetLifetimePnL returned NaN")
	}
}

func TestTxHashesUnmarshalled(t *testing.T) {
	trades, err := pnllogger.GetRecentTrades(10)
	if err != nil {
		t.Fatalf("GetRecentTrades: %v", err)
	}
	for _, tr := range trades {
		if tr.TxHashes == nil {
			t.Errorf("trade %s has nil TxHashes (expected slice)", tr.ID)
		}
	}
}

func TestPartialTradeMarkedForRecovery(t *testing.T) {
	buyTx := "partial-buy-tx"
	partial := types.TradeResult{
		TradeID:   fmt.Sprintf("test-%d-partial", time.Now().UnixNano()),
		Status:    types.TradePartial,
		BuyTx:     &buyTx,
		ProfitUSD: 0,
		FeesUSD:   0.5,
		NetPnLUSD: -0.5,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}
	if err := pnllogger.LogTrade(mockOpportunity, partial, 100); err != nil {
		t.Fatalf("LogTrade (partial): %v", err)
	}
	recoveries, err := pnllogger.GetTradesNeedingRecovery()
	if err != nil {
		t.Fatalf("GetTradesNeedingRecovery: %v", err)
	}
	if len(recoveries) == 0 {
		t.Error("expected at least one recovery trade after logging a partial trade")
	}
}
